### ISSUE TITLE
Unexpected type modification when storing issubclass check in bool

### ISSUE THREAD

**Issue by 502E532E** [microsoft/pyright#10476]
**Describe the bug**
When I assign the result of `issubclass` to a bool, the type gets incorrectly modified. This only happens if the check is combined with a logical `and`, not when assigning directly. See my example code below.

While technically correct, it is really unexpected and produces warnings regarding `Unknown` where they are not neccessary. Mypy behaves as expected and does not modify the type.

**Code or Screenshots**
```python
from typing import reveal_type

class Foo: ...

def test(bar: type):
    reveal_type(bar)  # type
    some_bool = True and issubclass(bar, Foo)
    reveal_type(bar)  # type[Foo] | type[Unknown]
```

**VS Code extension or command-line**
Tested in [pyright-playground](https://pyright-play.net/?code=MYGwhgzhAEBiD28Bc0B06BQGAmBTAZtAC64REAUARmAE4pECeADrgJRIbRfQ24BuuMCAD6jFlVqtoAYmLNcnbhHgBbXMMqIQ0ALzQAKjQCuuaGAB22aAEsoRyqEgQJNADRxErRV14Cho%2BRcpWTFcAG0EeABdaAAfORYwgFVzAGtzeAB3cyigA).

**Comment by erictraut** [comment]
Pyright is working as intended here, and I don't consider this a bug.

Pyright isn't "modifying the type". It's narrowing the type. When two branches of code flow merge, the narrowed types of an expression must be combined. Mypy chooses to combine them using a "join" operation, and pyright chooses to use a "union" operation. For more information, refer to [this documentation](https://microsoft.github.io/pyright/#/mypy-comparison?id=unions-vs-joins). A union operation preserves more type information and is almost always preferable to a join. One could argue that it's not preferable in your use case. Mypy has been changing most of their logic from joins to unions over time, so it's likely that you'll see the same behavior from mypy in a future version.
